### PR TITLE
Fixes for parking stall sampling with high availability

### LINKED ISSUE #3404: Long walk access-to-parking trips
Even in situations when there is unlimited parking availability, we're still ending up with some situations where we have long (sometimes very long) distances between where an agent wants to park and where the stall gets created, leading to some too-long walk access legs

### PR DESCRIPTION
**Pull request by zneedell**
A hypothesis for why we sometimes see really long walk access legs even when there is a lot of parking available:

This adjustment https://github.com/LBNL-UCB-STI/beam/blob/147029951f903d16b55052558295aaa8b7f4e9e0/src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala#L32 appears to be added to account for what happens when a request is near the edge of a zone. But, it assumes that zones are roughly circular, which isn't always the case. In cases with really elongated zones, we can end up with a situation where a request falls outside of the `tazCharacteristicDiameter` of the closest TAZ center. This would mean that the parking stall sampling method would not be able to create a stall at the location of the request, even if availability were infinite. In some cases, this could mean long walk access distances even when there are a lot of available stalls.

This fix has two components. 1: Passing a `closestZone` parameter to `availabilityAwareSampling` that ensures that the sampling distribution for the parking stall location is centered on the request, 2: Setting a maximum access distance for a parking stall within the closest zone to a request (for now this is hardcoded but should probably be configurable).

Might address #3394

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3405)

### CODE REVIEW

**Review by haitamlaarabi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3405)* status: 0 of 7 files reviewed, 2 unresolved discussions (waiting on @nikolayilyin and @zneedell)

___
*[src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala, line 41 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3405#-MqHGH1p3ML5INYbui9A:-MqHGH1q9sl8fMOosmaD:bn795c7) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/12758c740ad4bff29761e5ff03048cd6244d3f85/src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala#L41)):*
> ```Scala
>     // might be placed from the agent's desired destination
>     val availabilityFactor: Double =
>       if (availabilityRatio < Math.exp(-4.0)) 1.0 else -0.25 * math.log(availabilityRatio)
> ```

what exp(-4.0) stands for ?
___
*[src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala, line 55 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3405#-MqHFemA22wE2L4H66sA:-MqHFemA22wE2L4H66sB:b-lmhpz9) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/12758c740ad4bff29761e5ff03048cd6244d3f85/src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala#L55)):*
> ```
> taz.coord.getY - xDistance
> ```
I think you meant: taz.coord.getY - yDistance

**Review by zneedell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3405)* status: 0 of 7 files reviewed, 2 unresolved discussions (waiting on @haitamlaarabi and @nikolayilyin)

___
*[src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala, line 41 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3405#-MqHGH1p3ML5INYbui9A:-MqHML5u5o4_G0dFUPDU:bte7b6p) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/12758c740ad4bff29761e5ff03048cd6244d3f85/src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala#L41)):*
<details><summary><i>Previously, haitamlaarabi (Haitam Laarabi) wrote…</i></summary><blockquote>

what exp(-4.0) stands for ?
</blockquote></details>

It's the value for availabilityRatio for which -0.25*log(availabilityRatio) = 1. Not sure exactly where that function comes from, but it keeps there from being a discontinuity there and caps availabilityFactor at 1
___
*[src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala, line 55 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3405#-MqHFemA22wE2L4H66sA:-MqHM7ik3hOUzNipzPC8:bbvl47x) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/12758c740ad4bff29761e5ff03048cd6244d3f85/src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala#L55)):*
<details><summary><i>Previously, haitamlaarabi (Haitam Laarabi) wrote…</i></summary><blockquote>

> ```
> taz.coord.getY - xDistance
> ```
I think you meant: taz.coord.getY - yDistance
</blockquote></details>

Ah yep, thanks

File: src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala [hunk 1]
```diff
@@ -162,6 +162,7 @@ class ChargingFunctions[GEO: GeoLevel](
   override protected def sampleParkingStallLocation(
     inquiry: ParkingInquiry,
     parkingZone: ParkingZone[GEO],
-    geoArea: GEO
-  ): Coord = super[ParkingFunctions].sampleParkingStallLocation(inquiry, parkingZone, geoArea)
+    geoArea: GEO,
+    inClosetZone: Boolean = false
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala]: Typo: Closest

Comment (JustinPihony) [reply src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala]: This propagates throughout, so consider this a comment for the other typos.

**Comment by zneedell** [comment]
Test!

**Review by JustinPihony — APPROVED** [review]
Minor item

**Comment by JustinPihony** [comment]
Seems to be a stuck agent....then these failed tests (possibly related):

> 01:30:05 HierarchicalParkingManagerSpec:
> 01:30:06 HierarchicalParkingManager with no taz
> 01:30:06 - should return a response with an emergency stall *** FAILED *** (2 milliseconds)
> 01:30:06   java.lang.NullPointerException:
> 01:30:06   at beam.agentsim.infrastructure.parking.GeoLevel$$anon$1.getId(GeoLevel.scala:34)
> 01:30:06   at beam.agentsim.infrastructure.parking.GeoLevel$$anon$1.getId(GeoLevel.scala:33)
> 01:30:06   at beam.agentsim.infrastructure.InfrastructureFunctions.searchForParkingStall(InfrastructureFunctions.scala:129)
> 01:30:06   at beam.agentsim.infrastructure.HierarchicalParkingManager.processParkingInquiry(HierarchicalParkingManager.scala:84)
> 01:30:06   at beam.agentsim.infrastructure.HierarchicalParkingManagerSpec.$anonfun$new$7(HierarchicalParkingManagerSpec.scala:140)
> 01:30:06   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
> 01:30:06   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
> 01:30:06   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
> 01:30:06   at org.scalatest.Transformer.apply(Transformer.scala:22)
> 01:30:06   at org.scalatest.Transformer.apply(Transformer.scala:20)
> 01:30:06   ...



> 01:37:49 ParallelParkingManagerSpec:
> 01:37:49 ParallelParkingManager with no taz
> 01:37:49 - should return a response with an emergency stall *** FAILED *** (0 milliseconds)
> 01:37:49   java.lang.NullPointerException:
> 01:37:49   at beam.agentsim.infrastructure.parking.GeoLevel$$anon$1.getId(GeoLevel.scala:34)
> 01:37:49   at beam.agentsim.infrastructure.parking.GeoLevel$$anon$1.getId(GeoLevel.scala:33)
> 01:37:49   at beam.agentsim.infrastructure.InfrastructureFunctions.searchForParkingStall(InfrastructureFunctions.scala:129)
> 01:37:49   at beam.agentsim.infrastructure.parking.ParkingNetwork.$anonfun$processParkingInquiry$1(ParkingNetwork.scala:29)
> 01:37:49   at scala.Option.map(Option.scala:230)
> 01:37:49   at beam.agentsim.infrastructure.parking.ParkingNetwork.processParkingInquiry(ParkingNetwork.scala:29)
> 01:37:49   at beam.agentsim.infrastructure.ParallelParkingManager.processParkingInquiry(ParallelParkingManager.scala:103)
> 01:37:49   at beam.agentsim.infrastructure.ParallelParkingManagerSpec.$anonfun$new$7(ParallelParkingManagerSpec.scala:131)
> 01:37:49   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
> 01:37:49   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
> 01:37:49   ...


> 01:37:49 GHRouterSpec:
> 01:37:49 Static GH
> 01:37:49 - must add alternative route for GraphHopper if enabled *** FAILED *** (58 seconds, 738 milliseconds)
> 01:37:49   false was not equal to true (GHRouterSpec.scala:50)

**Comment by zneedell** [comment]
Test!

**Review by haitamlaarabi — APPROVED** [review]
> Reviewed 2 of 7 files at r1, 1 of 1 files at r4, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3405)* status: all files reviewed, 3 unresolved discussions (waiting on @haitamlaarabi, @nikolayilyin, and @zneedell)

**Comment by haitamlaarabi** [comment]
Test now!

**Pull request merged**